### PR TITLE
Introduce ZillowHousing model and show it in the detail page

### LINKED ISSUE #166: Zillow model
- [ ] Store Zillow data in our database
- [ ] OneToManyRelationship Location <-> Zillow
- [ ] When user visits a detail page and if there is no zillow data associated with the Location, trigger fetching zillow data and store it to database

### PR DESCRIPTION
**Pull request by alldne**
- closes #166 
- stores zillow data in our database as cache but we don't have expiration date for it yet
- shows zillow data to the detail page. To test, go `/location/6`

### CODE REVIEW

**Comment by alldne** [comment]
You need ZWDID to run it on you local

https://www.zillow.com/howto/api/APIOverview.htm
```
export ZWSID="<put your zwsid here>"
```
And run `python manage.py runserver`

File: external/cache/zillow.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        loc_addr = addr_parser.parse(location.full_address)
+        response_addr = addr_parser.parse(response.address.full_address)
+        if loc_addr["street_full"] != response_addr["street_full"]:
+            continue
+        if location.city != response.address.city:
+            continue
+        if location.state != response.address.state:
+            continue
+
+        ZillowHousing.objects.create(
```
Comment (ab7289) [inline external/cache/zillow.py]: it looks like this is only creating new zillow records unless I missed something? I think if we are truly just caching this data, then we need to delete the existing data when we determine that it needs to be refreshed.

File: external/zillow/tests.py [hunk 2]
```diff
@@ -21,4 +25,13 @@ def test_get_zillow_housing(self):
...
+            self.assertTrue(isinstance(housing, ZillowHousingResponse))
+
+    @mock.patch("external.zillow.fetch.get_zws_id", get_zws_id_stub)
+    @mock.patch("external.zillow.fetch.requests")
+    def test_get_zillow_housing_no_match_results(self, mock_requests):
+        mock_requests.get().content = get_zillow_error_response()
+        results = get_zillow_housing(
+            address="Jay St", zipcode="11201", show_rent_z_estimate=True
+        )
+        self.assertEqual(len(results), 0)
```
Comment (KShih) [inline external/zillow/tests.py]: It would be nice to add comments about why should the result length is 0.

File: external/cache/zillow.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            last_estimated=response.last_estimated,
+            url=response.url,
+            suite_num=response_addr.get("suite_num", ""),
+            location=location,
+        )
+    location.last_fetched_zillow = timezone.now()
+    location.save()
+
+
+def refresh_zillow_housing_if_needed(location):
```
Comment (KShih) [inline external/cache/zillow.py]: 👍

File: location/views.py [hunk 4]
```diff
@@ -1,7 +1,13 @@
 from django.views import generic
 from .models import Location
+from external.cache.zillow import refresh_zillow_housing_if_needed
 
 
 class LocationView(generic.DetailView):
     model = Location
     template_name = "location.html"
+
+    def get_object(self):
```
Comment (KShih) [inline location/views.py]: May I ask when or where does this function take place in?

File: location/views.py [hunk 5]
```diff
@@ -1,7 +1,13 @@
...
+        obj = super().get_object()
```
Comment (KShih) [inline location/views.py]: 👍

File: location/templates/location.html [hunk 6]
```diff
@@ -23,4 +23,27 @@
...
+
+<h2>Zillow</h2>
+{% if object.zillow_set.all %}
+<ul>
+{% for zillow in object.zillow_set.all %}
+<li>
+  <a href={{zillow.url}}>
+  <h3>
+    {% if zillow.suite_num %}
+      <span class="badge badge-secondary">#{{zillow.suite_num}}</span>
```
Comment (mok232) [inline location/templates/location.html]: Nice!

File: external/cache/zillow.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+
+def refresh_zillow_housing(location):
+    results = get_zillow_housing(
+        address=location.address,
+        city_state=f"{location.city}, {location.state}",
+        zipcode=location.zipcode,
+    )
+
+    addr_parser = StreetAddressParser()
```
Comment (mok232) [inline external/cache/zillow.py]: 👍

File: external/models.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    location = models.ForeignKey(
+        Location, on_delete=models.CASCADE, related_name="zillow_set"
+    )
+
+    def __str__(self):
+        return f"ZillowHousing({self.zpid}, {self.url}) - {self.location.address}"
+
+    @property
+    def estimated_rent_price_for_display(self):
+        return f"{self.estimated_rent_price_currency} {self.estimated_rent_price}"
```
Comment (mok232) [inline external/models.py]: nice formatting!

File: external/cache/zillow.py (hunk 1 above)
Comment (alldne) [reply external/cache/zillow.py]: Ooooops I missed some important points. Thanks for pointing it out. Let me fix it

**Comment by alldne** [comment]
Please do not merge this PR until I resolve this comment:
https://github.com/gcivil-nyu-org/fall2019-cs-gy-6063-team-five_fries/pull/173/files/cb497698ba2d9e45887c81138ae15c337d25b64c#r342856566

File: external/cache/zillow.py (hunk 1 above)
Comment (alldne) [reply external/cache/zillow.py]: Fixed it

File: external/zillow/tests.py (hunk 2 above)
Comment (alldne) [reply external/zillow/tests.py]: Added a comment

**Pull request merged**